File: frontend/src/types/company.ts
export interface Company {
  id: number;
  name: string;
  code: string;
  logo_path: string | null;
  address: string | null;
  city: string | null;
  state: string | null;
  country: string | null;
  pincode: string | null;
  phone: string | null;
  email: string | null;
  website: string | null;
  gst_number: string | null;
  pan_number: string | null;
  timezone: string | null;
  currency: string | null;
  is_active: boolean;
  theme_config: Record<string, any> | null;
  payroll_config: Record<string, any> | null;
  attendance_config: Record<string, any> | null;
  features: Record<string, boolean> | null;
  created_at?: string;
  updated_at?: string;
}

export interface CompanyCreate {
  name: string;
  code: string;
  address?: string;
  city?: string;
  state?: string;
  country?: string;
  pincode?: string;
  phone?: string;
  email?: string;
  website?: string;
  gst_number?: string;
  pan_number?: string;
  timezone?: string;
  currency?: string;
}

export interface CompanyUpdate {
  name?: string;
  code?: string;
  address?: string;
  city?: string;
  state?: string;
  country?: string;
  pincode?: string;
  phone?: string;
  email?: string;
  website?: string;
  gst_number?: string;
  pan_number?: string;
  timezone?: string;
  currency?: string;
  is_active?: boolean;
  theme_config?: Record<string, any>;
  payroll_config?: Record<string, any>;
  attendance_config?: Record<string, any>;
  features?: Record<string, boolean>;
}

export interface CompanyStats {
  id: number;
  name: string;
  code: string;
  is_active: boolean;
  employee_count: number;
  user_count: number;
  admin_count: number;
  created_at?: string;
}

export interface Permission {
  id: number;
  code: string;
  name: string;
  module: string;
  description: string | null;
}

export interface Role {
  id: number;
  name: string;
  company_id: number | null;
  is_system: boolean;
  description: string | null;
  permissions: Permission[];
  created_at?: string;
}

export interface RoleCreate {
  name: string;
  company_id?: number;
  description?: string;
  permissions: number[];
}

export interface AuditLog {
  id: number;
  user_id: number | null;
  company_id: number | null;
  action: string;
  entity_type: string | null;
  entity_id: number | null;
  details: string | null;
  ip_address: string | null;
  created_at?: string;
}

export interface MasterOverview {
  total_companies: number;
  active_companies: number;
  total_employees: number;
  total_users: number;
  users_by_role: Record<string, number>;
  employees_by_company: { company: string; count: number }[];
  recent_audit_logs: AuditLog[];
}

File: backend/schemas/company.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime


class CompanyCreate(BaseModel):
    name:       str = Field(..., min_length=1, max_length=255)
    code:       str = Field(..., min_length=1, max_length=50)
    address:    Optional[str] = None
    city:       Optional[str] = None
    state:      Optional[str] = None
    country:    Optional[str] = "India"
    pincode:    Optional[str] = None
    phone:      Optional[str] = None
    email:      Optional[str] = None
    website:    Optional[str] = None
    gst_number: Optional[str] = None
    pan_number: Optional[str] = None
    timezone:   Optional[str] = "Asia/Kolkata"
    currency:   Optional[str] = "INR"


class CompanyUpdate(BaseModel):
    name:       Optional[str] = Field(None, max_length=255)
    code:       Optional[str] = Field(None, max_length=50)
    address:    Optional[str] = None
    city:       Optional[str] = None
    state:      Optional[str] = None
    country:    Optional[str] = None
    pincode:    Optional[str] = None
    phone:      Optional[str] = None
    email:      Optional[str] = None
    website:    Optional[str] = None
    gst_number: Optional[str] = None
    pan_number: Optional[str] = None
    timezone:   Optional[str] = None
    currency:   Optional[str] = None
    is_active:  Optional[bool] = None
    theme_config:     Optional[dict] = None
    payroll_config:   Optional[dict] = None
    attendance_config: Optional[dict] = None
    features:         Optional[dict] = None


class CompanyResponse(BaseModel):
    id:         int
    name:       str
    code:       str
    logo_path:  Optional[str] = None
    address:    Optional[str] = None
    city:       Optional[str] = None
    state:      Optional[str] = None
    country:    Optional[str] = None
    pincode:    Optional[str] = None
    phone:      Optional[str] = None
    email:      Optional[str] = None
    website:    Optional[str] = None
    gst_number: Optional[str] = None
    pan_number: Optional[str] = None
    timezone:   Optional[str] = None
    currency:   Optional[str] = None
    is_active:  bool = True
    theme_config:     Optional[dict] = None
    payroll_config:   Optional[dict] = None
    attendance_config: Optional[dict] = None
    features:         Optional[dict] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class CompanyStats(BaseModel):
    id:               int
    name:             str
    code:             str
    is_active:        bool
    employee_count:   int = 0
    user_count:       int = 0
    active_employees: int = 0
